src/parse_util.rs

```diff
@@ -1433,26 +1462,28 @@ fn detect_errors_in_backgrounded_job(
         return false;
     };
 
-    let mut errored = false;
     // Disallow background in the following cases:
     // foo & ; and bar
     // foo & ; or bar
     // if foo & ; end
     // while foo & ; end
-    let Some(job_conj) = job.parent().unwrap().as_job_conjunction() else {
+    let NodeRef::Branch(BranchRef::JobConjunction(job_conj)) = job.parent().unwrap().as_node()
+    else {
         return false;
     };
 
-    if job_conj.parent().unwrap().as_if_clause().is_some()
-        || job_conj.parent().unwrap().as_while_header().is_some()
+    let mut errored = false;
+
+    if let NodeRef::Branch(BranchRef::IfClause(_) | BranchRef::WhileHeader(_)) =
+        job_conj.parent().unwrap().as_node()
```

Comment:
maybe this is better written using `matches!`?

Reply:
I'll do you one better: that entire block can be a `match`, and we don't need an `errored` variable.